osu.Game.Rulesets.Osu/Difficulty/Interp/HermiteSpline.cs

```diff
@@ -0,0 +1,41 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+namespace osu.Game.Rulesets.Osu.Difficulty.Interp
+{
+    public struct HermiteSpline
+    {
+        public HermiteSpline(double x0, double val0, double d0, double x1, double val1, double d1)
+        {
+            double scale = 1 / (x1 - x0);
+            double scale2 = scale * scale;
+
+            X0 = x0;
+            X1 = x1;
+            D1 = d1;
+            Val1 = val1;
+
+            C0 = val0;
+            C1 = d0;
+            C2 = 3 * (val1 - val0) * scale2 - (2 * d0 + d1) * scale;
+            C3 = (2 * (val0 - val1) * scale + d0 + d1) * scale2;
```

Comment:
are you sure this is correct? I mainly mean the `scale` part - I assume you're adding it to rescale *x* values onto the unit interval, but if that's the case, shouldn't `C1` be scaled by `scale`, `C2` by `scale2` entirely, and `C3` by `scale3` (which doesn't exist yet) entirely?

Reply:
I'm fairly sure it's right. 

x values are translated from (x1,x2) to (0,x2-x1), but not scaled down to (0,1). I just cross referenced against scipy [here](https://github.com/scipy/scipy/blob/v1.5.4/scipy/interpolate/_cubic.py#L136), `1/dxr` is `scale`, `slope` is `(val1-val0)*scale`

Reply:
nvm, it's much easier to cross reference it [here](https://github.com/mathnet/mathnet-numerics/blob/8cdd91d37f36cb4daecc7838898a08ef9c3aa1a4/src/Numerics/Interpolation/CubicSpline.cs#L77) there's less stuff factored out.

Reply:
Okay, after staring at this for a long time I think I get it. It's not that `x` values are not scaled down, it's that derivatives are *scaled back up* by `1 / scale` again.

Reply:
Just checked the refactor, you've changed `3*(...) * scaleSquared - 2*(...)*scale` to `(3*(...)* scale -2*(...))*scale`. Should still work. I personally would prefer not to nest brackets like that when it's not necessary, but it shouldn't change any results. 

If there are discrepancies in the refactor they are elsewhere.

Reply:
i mean it's a bit strange that c2 doesn't have nested brackets but c3 does, but sure, i guess